jq255.md

```diff
@@ -0,0 +1,1412 @@
+    *non-negative* if `sgn(x) = 0`.
+
+  - The elliptic curve is the set of points `(e, u)`, with both
+    coordinates being elements of `GF(q)` that fulfill the curve
+    equation `e^2 = (a^2 - 4*b)*u^4 - 2*a*u^2 + 1`, for two constants
+    `a` and `b` that are part of the curve definition. The curve neutral
+    point (the "point-at-infinity" which is not actually "at infinity"
+    in this coordinate system) is `I = (1,0)`. The curve also contains a
+    single point of order 2 denoted `N = (-1,0)`.
+    
+  - The underlying elliptic curve contains exactly `n = 2*r` points,
+    with `r` being a prime integer. The group itself (jq255e or jq255s)
+    has order `r`. For both groups, `r` is close to `2^254`.
+
+  - For ease of notation, we define `a' = -2*a` and `b' = a^2 - 4*b`.
+    With these constants, the curve equation becomes:
+    `e^2 = b'*u^4 + a'*u^2 + 1`
+
+  - A *group element* is defined as a pair `{ P, P+N }` for any curve
+    point `P`. The points `P` and `P+N` are the two possible
+    *representants* of the group element. The group law is denoted
+    additively: for any two group elements, their sum is computed by
+    simply adding together their representant points over the curve;
+    since `N` has order 2, it does not matter which representants are
+    used. The encoding process uses, for each group element, a specific
+    representant, and automatically computes it if given the other
+    representant, so that the encoded format is canonical. Similarly,
+    group element equality abstracts away the representants.
+
+  - Integers modulo `r` are called *scalars*. Since `r` is prime, the
+    scalars constitute a finite field. Group elements can be multiplied
+    by scalars: for a scalar `s` and a group element `P`, the product is
+    `s*P = P + P + ... + P`, where the group law is applied `si - 1`
+    times (with `si` being the integer representant of `s`, in the `0`
+    to `r-1` range).
+
+The jq255e curve parameters are the following:
+
+```
+jq255e:
+
+  q = 2^255 - 18651
+  a = 0
+  b = -2
+  a' = -2*a      = 0
+  b' = a^2 - 4*b = 8
+  equation: e^2 = 8*u^4 + 1
+  group order: r = 2^254 - 131528281291764213006042413802501683931
+  curve order: n = 2*r
+  conventional generator: G = (eg, ug)
+     eg = -3
+     ug = -1
+```
+
+For jq255s, the curve parameters are as follows:
+
+```
+jq255e:
+
+  q = 2^255 - 3957
+  a = -1
+  b = 1/2 = 2^254 - 1978
+  a' = -2*a      = 2
+  b' = a^2 - 4*b = -1
+  equation: e^2 = -u^4 + 2*u^2 + 1
+  group order: r = 2^254 + 56904135270672826811114353017034461895
+  curve order: n = 2*r
+  conventional generator: G = (eg, ug)
+     eg = 6929650852805837546485348833751579670837850621479164143703164723313568683024
+     ug = 3
+```
+
+## Group Element Operations
+
+In this section, practical procedures for computing point operations are
+presented using pseudocode in a syntax compatible with the Python language
+and the [SageMath] mathematics system. This syntax closely follows the
```

Comment:
It's unclear which procedures this refers to, as there's plenty of them using `^`.

Reply:
Actually it's all of them (or, if you prefer, all the ones which are in colour).

Reply:
I will detail here (I was not clear above): I use _mathematical notation_, in monospace font (with backquotes in the Markdown syntax), and I use _pseudocode_ (which is in fact real Python/Sage code), in code blocks (rendered in highlighted and coloured blocks). The pseudocode uses the `**` operator because Python somewhat requires it (the `^` operator being a XOR instead). Normally I did not use `**` outside of pseudocode. The `^` character can still occur in _code comments_ within code blocks.